### DISCUSSION TITLE
Let functions return instances of MediaWikiPage instead of titles

### DISCUSSION THREAD

**Discussion by Bomme** [barrust/mediawiki#118] [Ideas]
Hi, first of all thanks for your wonderful library!

I noticed the following:
When using methods like `search` or `categorymembers`, I need to make many repeated request to the API to get simple page information such as the `pageid`.
E.g. to get all page-ids in a category:
```python
>>> results = wikipedia.categorymembers("Python (programming language)", subcategories=False)
>>> type(results[0])
<class 'str'>
>>> results
['Python (programming language)', 'ActivePython', 'Asynchronous Server Gateway Interface', 'CircuitPython', 'CodeSkulptor', 'Core Python Programming', 'Cython', 'Dive into Python', 'Django Girls', 'Docstring']
>>> pages = [wikipedia.page(r) for r in results]  # does a new request for every page
>>> pages
[<MediaWikiPage 'Python (programming language)'>, <MediaWikiPage 'ActiveState'>, <MediaWikiPage 'Asynchronous Server Gateway Interface'>, <MediaWikiPage 'CircuitPython'>, <MediaWikiPage 'CodeSkulptor'>, <MediaWikiPage 'Core Python Programming'>, <MediaWikiPage 'Cython'>, <MediaWikiPage 'Mark Pilgrim'>, <MediaWikiPage 'Django Girls'>, <MediaWikiPage 'Docstring'>]
>>> page_ids = [p.pageid for p in pages]
```

The returned lists of the methods only contain the titles as strings while the API response in the methods contain all required fields to construct a minimal `MediaWikiPage` object.

What do you think about letting the methods return instances of `MediaWikiPage` instead to reduce the number of requests to the API?

**Comment by barrust** [comment]
This would be a great idea! Any help in doing this would be greatly appreciated.

**Reply by Bomme** [reply]
I was playing around with the code a little bit and I think the easiest would be to introduce a new result type:

```python
@dataclass(frozen=True)
class ResponseResult:
    ns: int
    title: str 
    pageid: int
```

Since the return value of the methods will change it is a breaking change and will require to up the version to 0.8.0. 

I can draw up a PR with the changes if you agree with the proposal.

**Reply by barrust** [reply]
That would be great! I'd love to see it and see how it could be used to reduce calls to the API.
